File: src/python/statset.py

class StatSet(object):

    STAT_NAMES = (
        '_hp', '_mp', 'maxhp', 'maxmp',
        'att', 'mag', 'pres', 'mres',
        'level', 'exp', 'next',
        )

    def __init__(self, **kw):
        for name in self.STAT_NAMES:
            setattr(self, name, kw.get(name, 0))
        # special hack:
        if 'hp' in kw:
            self.hp = kw['hp']
        if 'mp' in kw:
            self.mp = kw['mp']

    def __getitem__(self, key):
        return self.__dict__[key]

    def __setitem__(self, key, value):
        if key in self.__dict__:
            self.__dict__[key] = value
        else:
            raise KeyError, key

    hp = property(lambda self: self._hp, lambda self, value: setattr(self, '_hp', max(0, min(self.maxhp, value))))
    mp = property(lambda self: self._mp, lambda self, value: setattr(self, '_mp', max(0, min(self.maxmp, value))))

    def clone(self):
        s = StatSet()
        for name in self.STAT_NAMES:
            setattr(s, name, getattr(self, name))
        return s

File: src/winter/statset.ts
export class StatSet {
    att: number = 0
    exp: number = 0
    level: number = 0
    mag: number = 0
    maxhp: number = 0
    maxmp: number = 0
    mres: number = 0
    next: number = 0
    pres: number = 0

    private _hp: number = 0
    private _mp: number = 0

    get hp(): number { return this._hp }
    set hp(v: number) {
        this._hp = Math.max(0, Math.min(this.maxhp, v))
    }

    get mp(): number { return this._mp }
    set mp(v: number) {
        this._mp = Math.max(0, Math.min(this.maxmp, v))
    }

    clone(): StatSet {
        const s = new StatSet()

        s._hp = this._hp
        s._mp = this._mp
        s.att = this.att
        s.exp = this.exp
        s.level = this.level
        s.mag = this.mag
        s.maxhp = this.maxhp
        s.maxmp = this.maxmp
        s.mres = this.mres
        s.next = this.next
        s.pres = this.pres

        return s
    }
}
